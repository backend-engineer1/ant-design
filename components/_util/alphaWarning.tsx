import canUseDom from 'rc-util/lib/Dom/canUseDom';
import * as React from 'react';
import Notification from '../notification';

// Alpha version warning, should remove when published
/* istanbul ignore next */
if (process.env.NODE_ENV !== 'production' && process.env.NODE_ENV !== 'test' && canUseDom()) {
  const icon: React.ReactNode = (
    <img
      alt="antd"
      style={{
        filter: ' brightness(100)',
        width: 40,
        transform: 'translateX(-10px)',
      }}
      src="https://gw.alipayobjects.com/zos/rmsportal/KDpgvguMpGfqaHPjicRK.svg"
    />
  );

  Notification.open({
    duration: 0,
    placement: 'top',
    message: <span style={{ color: '#FFF' }}>You are using dev version</span>,
    description: (
      <div>
        <p>Which is used for validating and may not same as final release version.</p>
        <p>DO NOT USE IN YOUR PRODUCTION!</p>
        <p style={{ marginTop: 8 }}>
          Ref:
          <a
            target="_blank"
            rel="noreferrer"
            style={{ marginInline: 4, color: '#FFF', textDecoration: 'underline' }}
            href="https://github.com/ant-design/ant-design/issues/33862#user-content-alpha-offset"
          >
            #33862 - What is in experimental?
          </a>
        </p>
      </div>
    ),
    icon,
    closeIcon: <span style={{ color: '#FFF' }}>X</span>,
    style: {
      background: '#CE5345',
      color: '#FFF',
    },
  });
}
