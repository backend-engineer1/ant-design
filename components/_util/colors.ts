export const PresetStatusColorTypes = [
  'success',
  'processing',
  'error',
  'default',
  'warning',
] as const;

export const PresetColorTypes = [
  'pink',
  'red',
  'yellow',
  'orange',
  'cyan',
  'green',
  'blue',
  'purple',
  'geekblue',
  'magenta',
  'volcano',
  'gold',
  'lime',
] as const;

export type PresetColorType = typeof PresetColorTypes[number];

export type PresetStatusColorType = typeof PresetStatusColorTypes[number];
