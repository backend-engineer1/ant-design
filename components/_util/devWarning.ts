import devWarning, { resetWarned } from 'rc-util/lib/warning';

export { resetWarned };

export default (valid: boolean, component: string, message: string): void => {
  devWarning(valid, `[antd: ${component}] ${message}`);

  // StrictMode will inject console which will not throw warning in React 17.
  if (process.env.NODE_ENV === 'test') {
    resetWarned();
  }
};
