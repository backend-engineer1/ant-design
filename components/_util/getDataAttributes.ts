export default function getDataAttributes(props: any) {
  return Object.keys(props).reduce((prev: any, key: string) => {
    if (key.substr(0, 5) === 'data-') {
      prev[key] = props[key];
    }
    return prev;
  }, {});
}
