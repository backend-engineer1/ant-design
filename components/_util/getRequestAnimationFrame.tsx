function requestAnimationFramePolyfill() {
  let lastTime = 0;
  return function(callback) {
    const currTime = new Date().getTime();
    const timeToCall = Math.max(0, 16 - (currTime - lastTime));
    const id = window.setTimeout(function() { callback(currTime + timeToCall); }, timeToCall);
    lastTime = currTime + timeToCall;
    return id;
  };
}

export default function getRequestAnimationFrame() {
  if (typeof window === 'undefined') {
    return () => {};
  }
  if (window.requestAnimationFrame) {
    return window.requestAnimationFrame;
  }
  const prefix = ['moz', 'ms', 'webkit'].filter(key => `${key}RequestAnimationFrame` in window)[0];
  return prefix
    ? window[`${prefix}RequestAnimationFrame`]
    : requestAnimationFramePolyfill();
}
