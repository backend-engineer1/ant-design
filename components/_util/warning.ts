import * as React from 'react';
import rcWarning, { resetWarned } from 'rc-util/lib/warning';

export { resetWarned };
export function noop() {}

type Warning = (valid: boolean, component: string, message?: string) => void;

// eslint-disable-next-line import/no-mutable-exports
let warning: Warning = noop;
if (process.env.NODE_ENV !== 'production') {
  warning = (valid, component, message) => {
    rcWarning(valid, `[antd: ${component}] ${message}`);

    // StrictMode will inject console which will not throw warning in React 17.
    if (process.env.NODE_ENV === 'test') {
      resetWarned();
    }
  };
}

type BaseTypeWarning = (
  valid: boolean,
  /**
   * - deprecated: Some API will be removed in future but still support now.
   * - usage: Some API usage is not correct.
   * - breaking: Breaking change like API is removed.
   */
  type: 'deprecated' | 'usage' | 'breaking',
  message?: string,
) => void;

type TypeWarning = BaseTypeWarning & {
  deprecated: (valid: boolean, oldProp: string, newProp: string, message?: string) => void;
};

export interface WarningContextProps {
  deprecated?: boolean;
}

export const WarningContext = React.createContext<WarningContextProps>({});

/**
 * This is a hook but we not named as `useWarning`
 * since this is only used in development.
 * We should always wrap this in `if (process.env.NODE_ENV !== 'production')` condition
 */
export const devUseWarning: (component: string) => TypeWarning =
  process.env.NODE_ENV !== 'production'
    ? (component: string) => {
        const { deprecated } = React.useContext(WarningContext);

        const typeWarning: TypeWarning = (valid, type, message) => {
          if (deprecated !== false || type !== 'deprecated') {
            warning(valid, component, message);
          }
        };

        typeWarning.deprecated = (valid, oldProp, newProp, message) => {
          typeWarning(
            valid,
            'deprecated',
            `\`${oldProp}\` is deprecated. Please use \`${newProp}\` instead.${
              message ? ` ${message}` : ''
            }`,
          );
        };

        return typeWarning;
      }
    : () => {
        const noopWarning: TypeWarning = () => {};

        noopWarning.deprecated = noop;

        return noopWarning;
      };

export default warning;
