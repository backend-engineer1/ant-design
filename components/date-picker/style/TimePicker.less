.@{calendar-prefix-cls}-time {
  // Change display order in DOM
  .@{calendar-prefix-cls}-input-wrap {
    direction: rtl;
    .@{calendar-prefix-cls}-date-input-wrap,
    .@{calendar-prefix-cls}-time-picker-wrap {
      direction: ltr;
      display: inline-block;
    }
    .@{timepicker-prefix-cls}-panel,
    .@{calendar-prefix-cls}-clear-btn {
      direction: ltr;
    }
    .@{calendar-prefix-cls}-time-picker-wrap {
      position: absolute;
      right: 0;
    }
  }
  .@{calendar-prefix-cls}-input,
  .@{timepicker-prefix-cls}-input {
    // .input;
  }
  .@{timepicker-prefix-cls}-input {
    display: none;
  }
  .@{calendar-prefix-cls}-input {
    padding-right: 21px;
  }
  .@{timepicker-prefix-cls}-panel {
    min-width: 168px;
  }
  .@{timepicker-prefix-cls}-icon {
    font-family: "anticon" !important;
    position: relative;
    content: "\E643";
    display: inline;
  }
}
