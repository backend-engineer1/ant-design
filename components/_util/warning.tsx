import warning from 'warning';

let warned: Record<string, boolean> = {};

export function resetWarned() {
  warned = {};
}

export default (valid: boolean, component: string, message: string): void => {
  if (!valid && !warned[message]) {
    warning(false, `[antd: ${component}] ${message}`);
    warned[message] = true;
  }
};
