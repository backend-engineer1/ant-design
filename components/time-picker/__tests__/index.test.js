import React from 'react';
import { mount } from 'enzyme';
import dayjs from 'dayjs';
import customParseFormat from 'dayjs/plugin/customParseFormat';
import TimePicker from '..';
import focusTest from '../../../tests/shared/focusTest';
import mountTest from '../../../tests/shared/mountTest';
import { resetWarned } from '../../_util/warning';
import rtlTest from '../../../tests/shared/rtlTest';

dayjs.extend(customParseFormat);

describe('TimePicker', () => {
  const errorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});

  afterEach(() => {
    errorSpy.mockReset();
  });

  afterAll(() => {
    errorSpy.mockRestore();
  });

  focusTest(TimePicker, { refFocus: true });
  mountTest(TimePicker);
  rtlTest(TimePicker);

  it('warning for addon', () => {
    resetWarned();
    const addon = () => (
      <button className="my-btn" type="button">
        OK
      </button>
    );
    const wrapper = mount(<TimePicker addon={addon} open />);

    expect(wrapper.find('.my-btn').length).toBeTruthy();
    expect(errorSpy).toHaveBeenCalledWith(
      'Warning: [antd: TimePicker] `addon` is deprecated. Please use `renderExtraFooter` instead.',
    );
  });

  it('not render clean icon when allowClear is false', () => {
    const wrapper = mount(
      <TimePicker defaultValue={dayjs('2000-01-01 00:00:00')} allowClear={false} />,
    );
    expect(wrapper.render()).toMatchSnapshot();
  });

  it('clearIcon should render correctly', () => {
    const clearIcon = <div className="test-clear-icon">test</div>;
    const wrapper = mount(<TimePicker clearIcon={clearIcon} />);
    expect(wrapper.find('Picker').last().prop('clearIcon')).toEqual(
      <div className="test-clear-icon">test</div>,
    );
  });

  it('prop locale should works', () => {
    const locale = {
      placeholder: 'Избери дата',
    };
    const wrapper = mount(
      <TimePicker defaultValue={dayjs('2000-01-01 00:00:00')} open locale={locale} />,
    );
    expect(wrapper.render()).toMatchSnapshot();
  });

  it('should pass popupClassName prop to Picker as dropdownClassName prop', () => {
    const popupClassName = 'myCustomClassName';
    const wrapper = mount(
      <TimePicker
        defaultOpenValue={dayjs('00:00:00', 'HH:mm:ss')}
        popupClassName={popupClassName}
      />,
    );
    expect(wrapper.find('Picker').last().prop('dropdownClassName')).toEqual(popupClassName);
  });

  it('should pass popupClassName prop to RangePicker as dropdownClassName prop', () => {
    const popupClassName = 'myCustomClassName';
    const wrapper = mount(
      <TimePicker.RangePicker
        defaultOpenValue={dayjs('00:00:00', 'HH:mm:ss')}
        popupClassName={popupClassName}
      />,
    );
    expect(wrapper.find('RangePicker').at(1).prop('dropdownClassName')).toEqual(popupClassName);
  });

  it('should support bordered', () => {
    const wrapper = mount(
      <TimePicker
        className="custom-class"
        defaultValue={dayjs('2000-01-01 00:00:00')}
        bordered={false}
      />,
    );
    expect(wrapper.render()).toMatchSnapshot();
  });
});
