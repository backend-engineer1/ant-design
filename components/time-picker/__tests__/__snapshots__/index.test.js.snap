exports[`TimePicker renders addon correctly 1`] = `
<div
  class="ant-time-picker-panel-addon">
  <button>
    Ok
  </button>
</div>
`;
